keps/sig-network/3015-prefer-same-node/README.md

```diff
@@ -0,0 +1,496 @@
+
+By "locking down" the meaning of `PreferClose`/`PreferSameZone`, we
+potentially limit the ability of future proxies to improve traffic
+routing on their own.
+
+However, actually _improving_ traffic routing requires the proxy to
+have more information than `TrafficDistribution` currently provides.
+(For example: does the user want to keep traffic in the same zone
+because it's faster, or because it's cheaper?)
+
+However, we are not likely to ever end up with a highly-detailed
+system for explaining the user's relative preferences for latency,
+bandwidth, cost, server load, etc, and many users have indicated that
+they would be very happy with a way to _just_ express "prefer same
+zone" without worrying that this might thwart some theoretical future
+proxy. (And of course, if it _does_ thwart some theoretical future
+proxy, they could just change the Service definition at that point.)
+
+Also, while these alternative traffic distribution modes risk
+overloading endpoints if used poorly, the API does not need to assume
+that the service proxy is the _only_ component in the system that
+cares about balancing endpoint load; the user may be using other APIs
+such as Pod Topology Spread Constraints or Horizontal Autoscaling to
+ensure that load is handled appropriately, and in those scenarios, the
+alternatively traffic distribution modes may be completely safe.
+
+## Design Details
+
+### Renaming/Deprecation of `PreferClose`
+
+KEP-4444 defines `PreferClose` as:
+
+> * `PreferClose`: Indicates a preference for routing traffic to endpoints that
+>   are topologically proximate to the client. The interpretation of
+>   "topologically proximate" may vary across implementations and could encompass
+>   endpoints within the same node, rack, zone, or even region.
+
+We will add a new `TrafficDistribution` value, `PreferSameZone`, which
+is defined as follows:
+
+* `PreferSameZone`: Indicates a preference for routing traffic to
+  endpoints that are in the same zone as the client. In general, the
+  proxy should always route to a same-zone endpoint if any is
+  available.
+
+And we will say that henceforth, `PreferClose` is just a deprecated
+alias for `PreferSameZone`, and never means anything other than
+"prefer same zone".
+
+Note that this does not change the existing implementation of
+`PreferClose` at all, it only removes the possibility of future
+changes.
+
+(In theory, service proxies other than kube-proxy may already have
+been implementing `PreferClose` with semantics other than the ones
+kube-proxy used. Those implementations will need to be updated when
+they add support for this KEP.)
+
+### Addition of `PreferSameNode`
+
+We will add a new `TrafficDistribution` value, `PreferSameNode`, which
+is defined as follows:
+
+* `PreferSameNode`: Indicates a preference for routing traffic to
+  endpoints that are on the same node as the client. In general, the
+  proxy should always route to a same-node endpoint if any is
+  available.
```

Comment:
How does this field interact with the iTP field? 

I assume that what is written here will also apply to `PreferSameNode`: https://kubernetes.io/docs/reference/networking/virtual-ips/#interaction-with-traffic-policies

Reply:
Yes. All topology-ish features should behave the same way with respect to other features; the only difference is which endpoints they pick as "topologically available".

`Local` traffic policy ignores topology, because its own routing concerns render topology irrelevant. But if we added other traffic policy types in the future, they might work differently and we'd have to define that then.